Strapi v4 compatibility + Unittests +  other features - another prof of concept

bluszcz opened the pull request:
This PR contains several changes:

* Basic compatibility with Strapi 4 (Collections are working, Single Type not yet) #8 
* Authentication (tested with Content API Token and Personal Token) #16 
* Filter to fetch media object types from the Strapi 4 Instance
* Better error handling (distinguish 401 from 403, and handle 404 more accurate) #6 
* Extra custom parameters - I believe it solves #15
* Unttests with 'mock/fixtures' - Executed in GH Actions
* "DevOps" - packages repository (GitHub Packages + RubyGems)  + GH Actions
* New Package Name - *jekyll-strapi-4* to avoid breaking working installations with Strapi 3 using `jekyll-strapi` older version of the plugin
* Locales - WIP - currently permalinks are supported.
* Adds mascot of the project - *Deerus*.

When comparing to my previously PRs:

* Improved code and docs
* Better understanding structure and API
* Fixed bug when copying assets was not happening executing filter

TODO:

* There are still some incorrect parameters of urls to rewrite
* SingleType
* Pagination

Currently working on 'meta' and 'error' parts of the response, which would allow me soon implement new things like pagination etc.

In long term squash would be necessary for commits, but for alpha review it should be fine...

bluszcz committed:
#22 #20 Introduces compatibility with Strapi
Keeps basic functionality of permalinks
Adds new filter *asset_url*

bluszcz committed:
Fix the table

bluszcz committed:
New table

bluszcz committed:
Replace with a correct name of a type

bluszcz committed:
Added authentication

bluszcz committed:
Fix typo

bluszcz committed:
Improved logging

bluszcz committed:
Remove not needed document.title
Create assets dir if it does not exist in first place

bluszcz committed:
empty? replaced with nil check

bluszcz committed:
Better logic for HTTP requests

bluszcz committed:
Better handlind HTTP response status codes

oDinZu committed:
add .env to .gitignore

We don't want to push secret data to git server.

bluszcz committed:
Build without warnings and errors

bluszcz committed:
[DEV] More docs in the code
[FIX] Use of StaticFile to ensure copying of assets
[DEV] Improved HTTP requests
[DOC] Updated README

bluszcz committed:
Remove spare space

csharpee commented:
I will do testing next time I am available.

Awesome 😎 @bluszcz

bluszcz commented:
@csharpee thank you for the tests :)

oDinZu committed:
Duplicate LICENSE

Only one LICENSE is needed.

csharpee commented on lib/jekyll/strapi/collection.rb:
```diff
@@ -20,28 +20,21 @@ def generate?
         end
-
-        result.each {|x| yield(x)}
+        data.each {|x| yield(x)}
```
yay!

csharpee commented:
On testing the following data in my photos.html layout: 

```<img src="{{ page.document.strapi_attributes.Image.data.attributes.formats.thumbnail| asset_url }}"/>```

Then, on Jekyll build, it errors on downloading any image data.

```
-jekyll-serve-1  |  Jekyll StrapiHTTP: Fetching entries from https://api.example.com/api/photos/2?populate=* using headers: ["Authorization"]
-jekyll-serve-1  |   Liquid Exception: undefined method `[]' for nil:NilClass uri_path = "#{strapi_endpoint}#{input['url']}" ^^^^^^^ in /srv/jekyll/_layouts/photos.html
-jekyll-serve-1  | /usr/gem/gems/jekyll-strapi-0.1.3/lib/jekyll/tags/strapiimagefilter.rb:22:in `asset_url': undefined method `[]' for nil:NilClass (NoMethodError)
-jekyll-serve-1  | 
-jekyll-serve-1  |         uri_path = "#{strapi_endpoint}#{input['url']}"
-jekyll-serve-1  |                                              ^^^^^^^
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/strainer.rb:56:in `invoke'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/context.rb:86:in `invoke'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/variable.rb:84:in `block in render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/variable.rb:82:in `each'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/variable.rb:82:in `inject'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/variable.rb:82:in `render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:103:in `render_node_to_output'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:80:in `render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/for.rb:160:in `block (2 levels) in render_segment'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/for.rb:158:in `each'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/for.rb:158:in `block in render_segment'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/context.rb:123:in `stack'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/for.rb:150:in `render_segment'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/for.rb:79:in `render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:103:in `render_node_to_output'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:82:in `render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/if.rb:46:in `block (2 levels) in render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/if.rb:44:in `each'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/if.rb:44:in `block in render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/context.rb:123:in `stack'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/tags/if.rb:43:in `render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:103:in `render_node_to_output'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:82:in `render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:208:in `block in render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:242:in `with_profiling'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:207:in `render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:220:in `render!'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:39:in `block (3 levels) in render!'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:59:in `measure_counts'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:38:in `block (2 levels) in render!'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:63:in `measure_bytes'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:37:in `block in render!'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:70:in `measure_time'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:36:in `render!'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/renderer.rb:131:in `render_liquid'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/renderer.rb:194:in `render_layout'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/renderer.rb:163:in `place_in_layouts'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/renderer.rb:93:in `render_document'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/renderer.rb:63:in `run'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:547:in `render_regenerated'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:539:in `block in render_pages'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:538:in `each'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:538:in `render_pages'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:211:in `render'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:80:in `process'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:28:in `process_site'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/commands/build.rb:65:in `build'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/commands/build.rb:36:in `process'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `block in process_with_graceful_fail'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `each'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `process_with_graceful_fail'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/commands/serve.rb:86:in `block (2 levels) in init_with_program'
-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `block in execute'
-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `each'
-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `execute'
-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/program.rb:44:in `go'
-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary.rb:21:in `program'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/exe/jekyll:15:in `<top (required)>'
-jekyll-serve-1  | 	from /usr/local/bundle/bin/jekyll:27:in `load'
-jekyll-serve-1  | 	from /usr/local/bundle/bin/jekyll:27:in `<main>'
-jekyll-serve-1 exited with code 1
```

csharpee commented:
On another note, I think I figured out a solution without making it super complicated. 

The general idear is create a base64 code out of all the images, then decode the image into Jekyll from Strapi API. 

# HOWTO: 

1. Create _plugins folder in root of your site directory. 
2. Create new file called base64_filter.rb >> copy/paste the following:
```
require "base64"

module Base64Filter
  def base64_encode (input)
    Base64.encode64(input)
  end
  def base64_decode (input)
    Base64.decode64(input)
  end
end

Liquid::Template.register_filter(Base64Filter) # register filter globally
```
4. Install Strapi plugin "**Placeholder**"
5. Upload **new** images to create a base64 code 
6. Goto Jekyll and transport placeholder data from Strapi API to Jekyll:
```
{{ page.email | base64_encode }}

{{ page.content | base64_decode }}
```
7. Refresh your Jekyll site and see the encoded or decoded image or email.

Note: This would work, but the plugin needs to see in: ```/api/upload/files?``` This location isn't a *collection*.

bluszcz commented:
So, 

I checked using following from docs configuration:
![image](https://user-images.githubusercontent.com/221943/174246393-6a120c24-c263-4ba2-94c3-fa60a1c77d66.png)

and result from http://localhost:1337/api/photos/1?populate=* which is used by my api is following (includes url attribute), so I am not sure here you data structure. Are you sure you are using same data structure from the README.md so it fits the docs html examples?

```
{"data":
  { 
    "id":1,
    "attributes": {"createdAt":"2022-06-14T18:14:46.528Z","updatedAt":"2022-06-17T07:17:31.203Z","publishedAt":"2022-06-14T18:14:49.968Z","Title":"Some title","Comment":"This is nice comment","Image":{"data":{"id":1,"attributes":{"name":"IMG_1389.JPG","alternativeText":"IMG_1389.JPG","caption":"IMG_1389.JPG","width":6960,"height":4640,"formats":{"thumbnail":{"name":"thumbnail_IMG_1389.JPG","hash":"thumbnail_IMG_1389_5a056fb742","ext":".JPG","mime":"image/jpeg","path":null,"width":234,"height":156,"size":3.88,"url":"/uploads/thumbnail_IMG_1389_5a056fb742.JPG"},"large":{"name":"large_IMG_1389.JPG","hash":"large_IMG_1389_5a056fb742","ext":".JPG","mime":"image/jpeg","path":null,"width":1000,"height":667,"size":27.72,"url":"/uploads/large_IMG_1389_5a056fb742.JPG"},"medium":{"name":"medium_IMG_1389.JPG","hash":"medium_IMG_1389_5a056fb742","ext":".JPG","mime":"image/jpeg","path":null,"width":750,"height":500,"size":18.27,"url":"/uploads/medium_IMG_1389_5a056fb742.JPG"},"small":{"name":"small_IMG_1389.JPG","hash":"small_IMG_1389_5a056fb742","ext":".JPG","mime":"image/jpeg","path":null,"width":500,"height":333,"size":10.58,"url":"/uploads/small_IMG_1389_5a056fb742.JPG"}},"hash":"IMG_1389_5a056fb742","ext":".JPG","mime":"image/jpeg","size":644.62,"url":"/uploads/IMG_1389_5a056fb742.JPG","previewUrl":null,"provider":"local","provider_metadata":null,"createdAt":"2022-06-14T19:05:14.226Z","updatedAt":"2022-06-14T19:05:14.226Z"}}}}},
"meta":{}}
```

bluszcz committed:
Merge pull request #3 from csharpee/patch-2

Duplicated LICENSE

bluszcz committed:
Added docs with example and diagram

Added info about deployed example

bluszcz committed:
Added docs with example and diagram

bluszcz committed:
Added info about deployed example

bluszcz committed:
Merge branch 'develop' of pi2.deernet:SSG/jekyll-strapi into develop

bluszcz committed:
Change Jekyll docs basename and url

bluszcz committed:
Remove welcome from index markdown

bluszcz committed:
Docs rewrite #1

bluszcz committed:
Docs rewrite #2

oDinZu commented on lib/jekyll/tags/strapiimagefilter.rb:
```diff
@@ -0,0 +1,41 @@
+      def asset_url(input)
+        assets_path = "assets" # TODO: to figure our in which conditions it can change
+        strapi_endpoint = @context.registers[:site].config['strapi']['endpoint']
+        uri_path = "#{strapi_endpoint}#{input['url']}"
```
Where is **url** being referenced? I don't see any other reference of url.

oDinZu replied:
https://api.example.com/api/photos/?populate=* is where the url is being referenced. I have this at:  ```"/uploads/quantum_qt5_ee7a744bdd.webp"```

oDinZu commented on lib/jekyll/tags/strapiimagefilter.rb:
```diff
@@ -0,0 +1,41 @@
+
+    module StrapiImageFilter
+      def asset_url(input)
+        assets_path = "assets" # TODO: to figure our in which conditions it can change
```
Is **assets_path** is this path specific to the _site folder or the root of the jekyll workspace? 

The _site folder for all my jekyll projects are: **assets** >> ***images*** >> *.webp

bluszcz replied:
Your default path "assets/images" should be transparent for filter - filter generates new url as return  of the function, so it should adapt... Perhaps I can add more debug in the filter!

csharpee commented:
here is my full trace log

```
-jekyll-serve-1  |   Liquid Exception: undefined method `[]' for nil:NilClass uri_path = "#{strapi_endpoint}#{input['url']}" ^^^^^^^ in pages/photos/index.html
-jekyll-serve-1  | /usr/gem/gems/jekyll-strapi-0.1.3/lib/jekyll/tags/strapiimagefilter.rb:25:in `asset_url': undefined method `[]' for nil:NilClass (NoMethodError)
-jekyll-serve-1  | 
-jekyll-serve-1  |         uri_path = "#{strapi_endpoint}#{input['url']}"
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/strainer.rb:56:in `invoke'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/context.rb:86:in `invoke'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/variable.rb:84:in `block in render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/variable.rb:82:in `each'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/variable.rb:82:in `inject'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/variable.rb:82:in `render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:103:in `render_node_to_output'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:80:in `render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:208:in `block in render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:242:in `with_profiling'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:207:in `render'
-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:220:in `render!'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:39:in `block (3 levels) in render!'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:59:in `measure_counts'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:38:in `block (2 levels) in render!'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:63:in `measure_bytes'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:37:in `block in render!'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:70:in `measure_time'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:36:in `render!'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/renderer.rb:131:in `render_liquid'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/renderer.rb:80:in `render_document'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/renderer.rb:63:in `run'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:547:in `render_regenerated'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:538:in `each'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:538:in `render_pages'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:211:in `render'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:80:in `process'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:28:in `process_site'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/commands/build.rb:65:in `build'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/commands/build.rb:36:in `process'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `block in process_with_graceful_fail'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `each'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `process_with_graceful_fail'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/commands/serve.rb:86:in `block (2 levels) in init_with_program'
-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `block in execute'
-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `each'
-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `execute'
-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/program.rb:44:in `go'
-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary.rb:21:in `program'
-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/exe/jekyll:15:in `<top (required)>'
-jekyll-serve-1  | 	from /usr/local/bundle/bin/jekyll:27:in `load'
-jekyll-serve-1  | 	from /usr/local/bundle/bin
```

bluszcz commented:
Can you show  me whole content `pages/photos/index.html` please?

oDinZu commented on lib/jekyll/strapi/strapihttp.rb:
```diff
@@ -0,0 +1,44 @@
+
+def strapi_request(url)
+    uri = URI(url)
+    req = Net::HTTP::Get.new(uri)
```
I wonder how difficult it would be to have `Down::NetHttps`

Docs: https://github.com/janko/down#downnethttp

oDinZu replied:
All the media files are located here: `https://api.example.com/api/upload/files?`

Do we have to use **collection: photo** to download images from the collection? I see how the collection could be used for something like a *gallery* for a single page, but a *gallery* is more of a feature, unless I am missing how this is being used?

bluszcz replied:
> I wonder how difficult it would be to have `Down::NetHttps`
> 
> Docs: https://github.com/janko/down#downnethttp

To be honest, not sure, I was not checking, but Down advertises itself as downloading utility... I haven't touched it here because it was working with default Ruby class, which I like. I used Down in other place because I didn't know how easily to download files and following some reading it was most robust solution without complex code.

csharpee commented on jekyll-strapi.gemspec:
```diff
@@ -10,13 +10,15 @@ Gem::Specification.new do |spec|
   spec.name = "jekyll-strapi"
   spec.license = "MIT"
-  spec.has_rdoc = false
+#  spec.has_rdoc = false
```
What does this change do?

csharpee commented:
This is 

> Can you show me whole content `pages/photos/index.html` please?

```
---
layout: photos
title: Photos
sub_title: A collection of articles about technology.
permalink: /photos/
---

---
layout: photos
title: Photos
sub_title: A collection of articles about technology.
permalink: /photos/
---
<div class="home">
  <h1 class="page-heading">{{ page.strapi_attributes.TestDescription }}</h1>
  <h2>{{ page.document.strapi_attributes.Title }}</h2>
  <p>{{ page.document.strapi_attributes.Comment }}</p>
  <img src="{{ page.document.strapi_attributes.Image.data.attributes.formats.thumbnail| asset_url }}"/>
</div>
```

I also tried `<img src="{{ photo.attributes.Image.data.attributes.formats.thumbnail.url }}"/>` and `

csharpee commented:
Ok, so in Graphql, this is the query I do to get all photos url.

```
query {
  photos {
    data {
      id
      attributes {
        Title
        Comment
        Image {
          data {
            attributes {
              url
            }
          }
        }
      }
    }
  }
}
```

bluszcz committed:
More debug into StrapiHTTP

bluszcz committed:
More debug can be now launch with:
bundle exec jekyll b --verbose

bluszcz commented:
> Update: This should work, but it doesn't `{{ photo.data.attributes.Image.data.attributes.url }}`
> 
> this doesn't work either: `{{ page.document.strapi_attributes.Image.data.attributes.url }}`
> 
> both tests produce **nil** or simply: **""** 👎
> 
> these work: 👍
> 
> ```
>       {{ photo.attributes.Comment }}
>       {{ photo.attributes.createdAt }}
> ```

TBH, all above is kind of unexpected for me in the context of the code being reviewed here :) I would like really to have simply verison of your Strapi --quickstart app (one fake entry of collection), and I could quickly debug and prepare working Jekyll-strapi template.

bluszcz committed:
Basic Rakefile for future tests

bluszcz committed:
Added setup and teardown to the example test
Added rdoc generation to Rakefile - it should motivate to write better documented code :)

bluszcz committed:
Simply rake for build

bluszcz committed:
Added rdoc to .gitignore

bluszcz committed:
First two source JSON fixtures data for the future unittests

bluszcz committed:
Pretty JSON in test files

csharpee commented:
On testing with your **/docs setup**, I received the exact same error. It may be related to HTTPS.

```
Jekyll StrapiHTTP: Fetching entries from https://api.example.com/api/photos/2?populate=* using headers: ["Authorization"]
docs-jekyll-serve-1  | StrapiImageFilter strapi_endpoint:https://api.example.com 
docs-jekyll-serve-1  | StrapiImageFilter input: 
docs-jekyll-serve-1  |   Liquid Exception: undefined method `[]' for nil:NilClass uri_path = "#{strapi_endpoint}#{input['url']}" ^^^^^^^ in docs.markdown
docs-jekyll-serve-1  | /usr/gem/gems/jekyll-strapi-0.1.3/lib/jekyll/tags/strapiimagefilter.rb:25:in `asset_url': undefined method `[]' for nil:NilClass (NoMethodError)
docs-jekyll-serve-1  | 
docs-jekyll-serve-1  |         uri_path = "#{strapi_endpoint}#{input['url']}"
docs-jekyll-serve-1  |                                              ^^^^^^^
docs-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/strainer.rb:56:in `invoke'
docs-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/context.rb:86:in `invoke'
docs-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/variable.rb:84:in `block in render'
docs-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/variable.rb:82:in `each'
docs-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/variable.rb:82:in `inject'
docs-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/variable.rb:82:in `render'
docs-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:103:in `render_node_to_output'
docs-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/block_body.rb:80:in `render'
docs-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:208:in `block in render'
docs-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:242:in `with_profiling'
docs-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:207:in `render'
docs-jekyll-serve-1  | 	from /usr/gem/gems/liquid-4.0.3/lib/liquid/template.rb:220:in `render!'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:39:in `block (3 levels) in render!'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:59:in `measure_counts'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:38:in `block (2 levels) in render!'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:63:in `measure_bytes'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:37:in `block in render!'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:70:in `measure_time'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/liquid_renderer/file.rb:36:in `render!'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/renderer.rb:131:in `render_liquid'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/renderer.rb:80:in `render_document'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/renderer.rb:63:in `run'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:547:in `render_regenerated'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:539:in `block in render_pages'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:538:in `each'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:538:in `render_pages'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:211:in `render'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:80:in `process'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:28:in `process_site'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/commands/build.rb:65:in `build'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/commands/build.rb:36:in `process'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `block in process_with_graceful_fail'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `each'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `process_with_graceful_fail'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/commands/serve.rb:86:in `block (2 levels) in init_with_program'
docs-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `block in execute'
docs-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `each'
docs-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `execute'
docs-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/program.rb:44:in `go'
docs-jekyll-serve-1  | 	from /usr/gem/gems/mercenary-0.4.0/lib/mercenary.rb:21:in `program'
docs-jekyll-serve-1  | 	from /usr/gem/gems/jekyll-4.2.2/exe/jekyll:15:in `<top (required)>'
docs-jekyll-serve-1  | 	from /usr/local/bundle/bin/jekyll:27:in `load'
docs-jekyll-serve-1  | 	from /usr/local/bundle/bin/jekyll:27:in `<main>'
docs-jekyll-serve-1 exited with code 1
```

bluszcz committed:
First unittest is working :)
Code is a mess - it was painful process going through Liquid
but it will be clean up soon.

bluszcz committed:
Fix indentation so the tests do not throw warning

bluszcz commented:
Added unittests.

bluszcz commented on test/test_collection.rb:
```diff
@@ -0,0 +1,104 @@
+
+    def test_create
+
+        @config_site = {"source"=>"/home/bluszcz/Dev/ssg/my-app-develop", "destination"=>"/home/bluszcz/Dev/ssg/my-app-develop/_site", "collections_dir"=>"", "cache_dir"=>".jekyll-cache", "plugins_dir"=>"_plugins", "layouts_dir"=>"_layouts", "data_dir"=>"_data", "includes_dir"=>"_includes", "collections"=>{"posts"=>{"output"=>true, "permalink"=>"/:categories/:year/:month/:day/:title:output_ext"}} ...
```
Yeah, this needs some clean up and more generic default values 🤦 😆

csharpee commented:
docs-jekyll-serve-1  | StrapiImageFilter 000 input: ==> YES 
docs-jekyll-serve-1  | StrapiImageFilter 111 input: 
docs-jekyll-serve-1  | StrapiImageFilter 222 context REGISTERS:{:site=>#<Jekyll::Site @source=/srv/jekyll>, :page=>{"layout"=>"photo", "content"=>"<div class=\"home\">\n  <h1 class=\"page-heading\">{{ page.strapi_attributes.TestDescription }}</h1>\n  <h2>{{ page.document.strapi_attributes.Title }}</h2>\n  <p>{{ page.document.strapi_attributes.Comment }}</p>\n  <img src=\"{{ page.document.strapi_attributes.Image.data.attributes.formats.thumbnail| asset_url }}\"/>\n</div>\n", "dir"=>"/", "excerpt"=>nil, "name"=>"photos.markdown", "path"=>"photos.markdown", "url"=>"/photos.html"}} 
docs-jekyll-serve-1  | StrapiImageFilter strapi_endpoint:https://api.example.com 
docs-jekyll-serve-1  |   Liquid Exception: undefined method `[]' for nil:NilClass uri_path = "#{strapi_endpoint}#{input['url']}" ^^^^^^^ in photos.markdown
docs-jekyll-serve-1  | /usr/gem/gems/jekyll-strapi-0.1.3/lib/jekyll/tags/strapiimagefilter.rb:28:in `asset_url': undefined method `[]' for nil:NilClass (NoMethodError)
docs-jekyll-serve-1  | 
docs-jekyll-serve-1  |       uri_path = "#{strapi_endpoint}#{input['url']}"

bluszcz commented:
So it seems not necessarily problem with filter per se, but in line StrapiImageFilter 111 input: we see that the StrapiDocumentDrop is not passed to the filter, which can point to some collection issue. 

Correct output would be something similar:

```
StrapiImageFilter 111 input: StrapiDocumentDrop
```
from the drops.rb.
I will try to add debug in other places. Thank you.

> ```
> docs-jekyll-serve-1  | StrapiImageFilter 000 input: ==> YES 
> docs-jekyll-serve-1  | StrapiImageFilter 111 input: 
> docs-jekyll-serve-1  | StrapiImageFilter 222 context REGISTERS:{:site=>#<Jekyll::Site @source=/srv/jekyll>, :page=>{"layout"=>"photo", "content"=>"<div class=\"home\">\n  <h1 class=\"page-heading\">{{ page.strapi_attributes.TestDescription }}</h1>\n  <h2>{{ page.document.strapi_attributes.Title }}</h2>\n  <p>{{ page.document.strapi_attributes.Comment }}</p>\n  <img src=\"{{ page.document.strapi_attributes.Image.data.attributes.formats.thumbnail| asset_url }}\"/>\n</div>\n", "dir"=>"/", "excerpt"=>nil, "name"=>"photos.markdown", "path"=>"photos.markdown", "url"=>"/photos.html"}} 
> docs-jekyll-serve-1  | StrapiImageFilter strapi_endpoint:https://api.example.com 
> docs-jekyll-serve-1  |   Liquid Exception: undefined method `[]' for nil:NilClass uri_path = "#{strapi_endpoint}#{input['url']}" ^^^^^^^ in photos.markdown
> docs-jekyll-serve-1  | /usr/gem/gems/jekyll-strapi-0.1.3/lib/jekyll/tags/strapiimagefilter.rb:28:in `asset_url': undefined method `[]' for nil:NilClass (NoMethodError)
> docs-jekyll-serve-1  | 
> docs-jekyll-serve-1  |       uri_path = "#{strapi_endpoint}#{input['url']}"
> ```

bluszcz committed:
Monkey patching for the UnitTests purpose. Evil :D

bluszcz committed:
Create ruby.yml #10

Pipeline to executing the tests

bluszcz committed:
Updated README.md

bluszcz committed:
#10 Gemfile
Updated authors

bluszcz committed:
Add test-unit to Gemfile #10

bluszcz committed:
#10 added "minima" theme

bluszcz committed:
#10 Changed directory in json files

bluszcz committed:
raw and endraw to the markdown files to ensure proper html examples generation

csharpee commented:
Currently, this jekyll-strapi plugin that is created doesn't work out of the box with Jekyll 4.2.2. If you fix that problem, the whole site structure doesn't work and the tutorials to guide other open source developers or users does NOT produce any successful results.

bluszcz commented:
@csharpee just for starter from your logs: ```https://api.example.com``` which shows that you area not using default configuration. Since you are not using default setup  from docs (its no https there) and you are unable to provide replayable environment - or clearly with code explain what  is - it impossible to  debug. This is an alpha patches extending first functionality, so you must first to ensure that  it works on setup configuration to understand how it works. Do not skip this step please ;) If  you will repeat an exercise and something does not work - you can just targz default strapi and jekyll projects and attach to the bug PR or make bug report. 

Other solution is that you write the uninttests describing your environment and failing case.

In the tests/ you can find how _config is being provision, you should be able to provision your own and write the simple test.

mkrajewski90 commented:
@bluszcz - I have some code changes to potential merge, should I make a fork & PR to your develop branch?

bluszcz commented:
Make pr please to my gh

mkrajewski90 committed:
target endpoint by defined collection type or collection name (default)

mkrajewski90 committed:
add parameters to requests

mkrajewski90 committed:
generate page by slug

mkrajewski90 committed:
possible to define populate

bluszcz commented:
Everyone interested - there are more changes incoming https://github.com/bluszcz/jekyll-strapi with the help of @mkrajewski90! <3 - check PR and comments there :)

mkrajewski90 committed:
add tests for url placeholders, endpoints, populate and path params

mkrajewski90 committed:
fixing warning in strapi page UT

bluszcz committed:
Merge pull request #21 from mkrajewski90/request-improvements

Request improvements

bluszcz commented:
There was a great PR from @mkrajewski90  which have been merged. 

If someone wants to simplify the tests I uploaded the package: https://github.com/bluszcz/jekyll-strapi/packages/1554094

bluszcz committed:
Test of the deploy action

bluszcz committed:
Remove GHA for a moment

bluszcz committed:
Create gem-push.yml

bluszcz committed:
Merge branch 'master' of github.com:bluszcz/jekyll-strapi

bluszcz committed:
Added changelog

bluszcz committed:
Build and push test

bluszcz committed:
Another test of the builds

bluszcz committed:
Version bump

bluszcz committed:
Create codeql-analysis.yml

bluszcz committed:
Merge branch 'master' of github.com:bluszcz/jekyll-strapi into develop

mkrajewski90 committed:
generate collection permalink

mkrajewski90 committed:
Permalinks on README

bluszcz committed:
Version bump - 1.0 hurray.

bluszcz committed:
Rebranding into jekyll-strapi-4

bluszcz committed:
Change name of gem

bluszcz committed:
Uncomment RubyGems

bluszcz committed:
Change branch name

bluszcz committed:
Update Gemfile

bluszcz committed:
Updating GH workflows

bluszcz commented:
https://rubygems.org/gems/jekyll-strapi-4 <- plugin have been released.

bluszcz committed:
Clean up of the package

bluszcz committed:
Added logo
Bumped version 1.0.1

bluszcz committed:
PNG typo

bluszcz committed:
Cute deer

bluszcz committed:
Extra line

bluszcz committed:
bump version

bluszcz committed:
Merge pull request #22 from mkrajewski90/collection-permalinks

generate collection permalink

bluszcz committed:
Merge branch 'develop' of github.com:bluszcz/jekyll-strapi-4 into develop

bluszcz committed:
Bump version

bluszcz committed:
Fixes #23
jekyll-strapi => jekyll-strapi-4

bluszcz committed:
Merge pull request #24 from bluszcz/23-cant-open-jekyllstrapi-classes

Fixes #23

bluszcz committed:
Bump version

bluszcz committed:
Bump version

bluszcz committed:
UT adapted to jekyll-strapi-4

bluszcz committed:
Specify min Ruby 3 -> http does not work on Ruby 2

bluszcz committed:
docs adaptation for strapi-4

bluszcz committed:
bump version

bluszcz committed:
Keep only ruby 3.0 for the tests in GH

bluszcz committed:
fixed action workflow file

bluszcz committed:
Change baseurl for GH pages

mkrajewski90 commented:
@bluszcz - do you want to merge develop or master? I see develop is not updated.

mkrajewski90 commented:
> > @bluszcz - do you want to merge develop or master? I see develop is not updated.
> 
> Tag current master as v3 or similar, and to master. What do you think?
> 
> Or you were asking about source branch from my repo?

I ask about source branch from your repo.
